src/async_impl/client.rs

```diff
@@ -1953,6 +1960,42 @@ impl ClientBuilder {
         self.config.quic_send_window = Some(value);
         self
     }
+
+    /// Adds a new Tower [`Layer`](https://docs.rs/tower/latest/tower/trait.Layer.html) to the
+    /// base connector [`Service`](https://docs.rs/tower/latest/tower/trait.Service.html) which
+    /// is responsible for connection establishment.a
+    ///
+    /// Each subsequent invocation of this function will wrap previous layers.
+    ///
+    /// If configured, the `connect_timeout` will be the outermost layer.
+    ///
+    /// Example usage:
+    /// ```
+    /// use std::time::Duration;
+    ///
+    /// # #[cfg(not(feature = "rustls-tls-no-provider"))]
+    /// let client = reqwest::Client::builder()
+    ///                      // resolved to outermost layer, meaning while we are waiting on concurrency limit
+    ///                      .connect_timeout(Duration::from_millis(200))
+    ///                      // underneath the concurrency check, so only after concurrency limit lets us through
+    ///                      .connector_layer(tower::timeout::TimeoutLayer::new(Duration::from_millis(50)))
+    ///                      .connector_layer(tower::limit::concurrency::ConcurrencyLimitLayer::new(2))
+    ///                      .build()
+    ///                      .unwrap();
+    /// ```
+    ///
+    pub fn connector_layer<L>(mut self, layer: L) -> ClientBuilder
+    where
+        L: Layer<BoxedConnectorService> + Clone + Send + Sync + 'static,
+        L::Service: Service<Uri, Response = Conn, Error = BoxError> + Clone + Send + Sync + 'static,
```

Comment:
This is perhaps a dumb idea, but I keep thinking about it, so may as well write it down:

Is there significant user benefit to be able to use (or even name) the input and output types of layers? (I think users can't even name the `Conn` type currently). What if this were `Service<Unnameable>`, and then users could not depend on it actually being a `Uri`? I'm sure use cases _exist_ that would want to access it, but are they significant?

The reason I've been thinking about this is that in order to support some other feature requests, like a per-request proxy, or UDS, or similar, it might eventually make sense to internally change the type from `Uri` to some `enum`. If users can name the type, that'd become a breaking change.

Reply:
Yeah, I can think of a few use cases. For instance allows selective handling such as applying different concurrency limits on connection establishment for different remote services. Or, you might want retries for some URIs but not others.

But, you can also accomplish the most important part of that by creating multiple clients, which wouldn't have overlapping connections to pool anyway.

An alternative that avoids breaking changes might be to instead make it a `Box<dyn Any>` which would allow the caller to downcast it to the various possible service inputs and still add whatever handling.

But, that a bit more friction for writing custom layers as well an extra allocation.

I'm also open to using `Service<Unnameable>` if we prefer simplicity.

What do you prefer?

Reply:
Hm, I don't think I'd want to box it...

Let's start with it being unnameable. That already allows for many generic layers to be added. If there's significant need to access the Uri in the future, we can either change it to Uri without breaking anyone, or by then tell people to use the Middleware to build up their own client stack. 🤷‍♂️

Reply:
Sounds good! Coming right up.

Reply:
Oh... we actually can't due to the bounds of the `Connect` trait that our underlying `hyper_util` client needs :/
`S: Service<Uri, Response = T> + Send + 'static,`

https://docs.rs/hyper-util/latest/hyper_util/client/legacy/connect/trait.Connect.html

Anyway I think we would run into trouble because we are supporting pre-1.74 rust versions where `private-in-public` is a hard error rather than lint warning.

Reply:
@seanmonstar how important is this to you? Do we care to plumb down to the `hyper_util` legacy client impl to modify the trait (probably across a breaking version boundary in case people are using `Uri`)?

My feeling is, probably we can just stick with it as-is and do the big cutover across breaking version boundaries for both packages if needed.

Reply:
> Connect trait that our underlying hyper_util client needs 

Do the layers themselves need to be `Connect`? We could do a tiny dance, where if you provide layers, inside `ConnectorService` we wrap it once more (not with an allocation) that wraps the argument in `Unnameable` and the layers are actually wrapped around a private layer that pulls it back out.

> private-in-public is a hard error rather than lint warning.

That's no big deal. I've gotten around that since Rust 1.0 by just fooling `mod stuff { pub mod internal { pub struct Unnameable; } }`.

Reply:
Actually had an idea on how to do it with a map request layer... taking a look. Though I think `private-in-public` being a compiler error on old versions might still block us.

Reply:
Yeah, this works:
```
pub(crate) struct Unnameable(Uri);

// So that the accepted layers are stored as:
pub(crate) type BoxedConnectorService = BoxCloneSyncService<Unnameable, Conn, BoxError>;

pub(crate) type BoxedConnectorLayer =
    BoxCloneSyncServiceLayer<BoxedConnectorService, Unnameable, Conn, BoxError>;

// And then the base service is wrapped:
 let nameable_service = ServiceBuilder::new().layer(MapRequestLayer::new(|request: Unnameable| request.0)).service(base_service);
 let mut service = BoxCloneSyncService::new(nameable_service);


// and service still keep the `Uri` request type
#[derive(Clone)]
pub(crate) enum Connector {
    Simple(ConnectorService),
    WithLayers(BoxCloneSyncService<Unnameable, Conn, BoxError>),
}

impl Service<Uri> for Connector {
...

impl Service<Uri> for ConnectorService {
...
```

Reply:
Nice trick, thanks for that. `Conn` and `Unnameable` are both crate-local now with no lint overrides needed. Let's see if CI works.

Reply:
@seanmonstar looks like CI/CD will pass, this approach worked. Thanks for the advice.